### PR TITLE
[zk-sdk] Add `errors`, `macros`, and `encryption` modules

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Summary of Changes
This is a follow-up to https://github.com/anza-xyz/agave/pull/988, which migrates the `errors`, `macros`, and `encryption` modules from `zk-token-sdk` into `zk-sdk`. The only additional change to these modules is that the deprecated functions are now removed.

Since the encryption module is largely the same in `zk-token-sdk` and `zk-sdk`, we can consider removing and re-exporting the functions/types in `zk-token-sdk` to the ones in `zk-sdk`. Since `zk-token-sdk` will not get any updates and then eventually be removed, I thought I would just leave it as is. We can consider re-exporting in a follow-up.

Fixes #

### CODE REVIEW

File: zk-sdk/src/macros.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
+// Internal macros, used to defined the 'main' impls
+#[macro_export]
```
Comment (joncinque) [inline zk-sdk/src/macros.rs]: Is the `#[macro_export]` annotation required?

**Review by joncinque — DISMISSED** [review]
Looks good! Just the one question

File: zk-sdk/src/macros.rs (hunk 1 above)
Comment (samkim-crypto) [reply zk-sdk/src/macros.rs]: Good point! I removed the annotation. Thanks!

**Comment by samkim-crypto** [comment]
I also realized that everything in the `macros` module is only required by the `elgamal` and `pedersen`, so I moved it inside `encryption` as well. I think this makes the crate organization a lot cleaner!

File: zk-sdk/src/encryption/mod.rs [hunk 2]
```diff
@@ -10,6 +10,9 @@
 //! - Basic type-wrapper around the AES-GCM-SIV symmetric authenticated encryption scheme
 //! implemented by [aes-gcm-siv](https://docs.rs/aes-gcm-siv/latest/aes_gcm_siv/) crate.
 
+#[cfg(not(target_os = "solana"))]
+#[macro_use]
```
Comment (joncinque) [inline zk-sdk/src/encryption/mod.rs:14]: Sorry, looking through this again, I don't think you need `#[macro_use]` either

**Review by joncinque — APPROVED** [review]
One last little thing, which you can do separately, looks good otherwise!

File: zk-sdk/src/encryption/mod.rs (hunk 2 above)
Comment (samkim-crypto) [reply zk-sdk/src/encryption/mod.rs:14]: Oh oops, yeah I guess this is only required by old rust. Thanks! I will removed on a follow-up.

Comment (samkim-crypto) [reply zk-sdk/src/encryption/mod.rs:14]: Hm, actually it seems like if I removed this `macro_use` then the macros seem to be not accessible to other crates. Reading a bit on the `macro_use` attribute, it seems like this is required here?

Comment (joncinque) [reply zk-sdk/src/encryption/mod.rs:14]: Ah ok, I thought the macros were only used internally. no worries then!

**Pull request merged**